### PR TITLE
Calculate CNV consequences based on supporting variants

### PR DESCRIPTION
**Pull request by nuno-agostinho**
[ENSVAR-5786](https://www.ebi.ac.uk/panda/jira/browse/ENSVAR-5786)

For CNVs with supporting evidence (basically those that we import from multiple sources), calculate the consequences using the supporting variants.

- If the supporting variants have `class_SO_term` as `deletion` or `loss`, treat CNV as a deletion.
- If the supporting variants have `class_SO_term` as `insertion`, `duplication` or `gain`, treat CNV as an insertion.

## Testing

From my understanding, VEP does not support using CNVs with supporting variants as input, so there is no way to test them via VEP. I opened ticket [ENSVAR-5793](https://www.ebi.ac.uk/panda/jira/browse/ENSVAR-5793) to support dbVar/DGVa identifiers, allowing to test this type of CNVs.

Check in my sandbox if consequences make sense for multiple CNVs with different supporting variants:
- CNV with deletions:
    - [nsv4517655](http://wp-np2-11.ebi.ac.uk:6070/Homo_sapiens/StructuralVariation/Mappings?db=core;r=1:521700-621900;sv=nsv4517655;svf=127710255;vdb=variation)
    - [nsv5594649](http://wp-np2-11.ebi.ac.uk:6070/Homo_sapiens/StructuralVariation/Mappings?db=core;r=13:100485511-101485511;sv=nsv5594649;svf=127234714;vdb=variation#)
- CNVs with duplications (considered as insertions):
    - [nsv4517661](http://wp-np2-11.ebi.ac.uk:6070/Homo_sapiens/StructuralVariation/Mappings?db=core;r=1:683134-783334;sv=nsv4517661;svf=127710249;vdb=variation)
    - [nsv4033399](http://wp-np2-11.ebi.ac.uk:6070/Homo_sapiens/StructuralVariation/Mappings?db=core;r=1:683134-783334;sv=nsv4033399;svf=127794743;vdb=variation)
    - [nsv6310981](http://wp-np2-11.ebi.ac.uk:6070/Homo_sapiens/StructuralVariation/Mappings?db=core;r=1:862572-962772;sv=nsv6310981;svf=130331903;vdb=variation)
- CNVs with copy number loss (considered as deletions):
    - [esv1812726](http://wp-np2-11.ebi.ac.uk:6070/Homo_sapiens/StructuralVariation/Mappings?r=1:196761405-196833442;sv=esv1812726;svf=1149079;vdb=variation)
    - [esv1815690](http://wp-np2-11.ebi.ac.uk:6070/Homo_sapiens/StructuralVariation/Mappings?r=5:177729898-177771948;sv=esv1815690;svf=1108895;vdb=variation)
    - [esv1850194](http://wp-np2-11.ebi.ac.uk:6070/Homo_sapiens/StructuralVariation/Mappings?r=3:98168147-98201019;sv=esv1850194;svf=1154027;vdb=variation)
- CNVs with copy number gain (considered as insertions):
    - [nsv443016](http://wp-np2-11.ebi.ac.uk:6070/Sus_scrofa/StructuralVariation/Mappings?r=7:67148496-67211415;sv=nsv443016;svf=665;vdb=variation)

### CODE REVIEW

File: modules/Bio/EnsEMBL/Variation/BaseVariationFeatureOverlapAllele.pm [hunk 1]
```diff
@@ -410,16 +410,27 @@ sub _bvf_preds {
...
   # use SO term to determine class
   if($is_sv) {
     my $class_SO_term = $bvf->class_SO_term;
 
-    if($class_SO_term =~ /deletion/) {
+    my $is_CNV = $class_SO_term eq "copy_number_variation";
+    my $is_CNV_deletion  = 0;
+    my $is_CNV_insertion = 0;
+    if ($is_CNV) {
+      my $support_vars  = $bvf->structural_variation->get_all_SupportingStructuralVariants;
```
Comment (dglemos) [inline modules/Bio/EnsEMBL/Variation/BaseVariationFeatureOverlapAllele.pm:422]: While testing VEP, the job failed when trying to fetch the supporting SVs.
Before fetching the supporting SVs you could check if `$bvf->structural_variation` is defined.

File: modules/Bio/EnsEMBL/Variation/Utils/VariationEffect.pm [hunk 2]
```diff
@@ -154,6 +154,17 @@ sub complete_overlap_feature {
     );
 }
 
+sub _supporting_cnv_terms {
+  # if variant is CNV, return class SO terms for its supporting variants
+  my $bvf = shift;
+
+  return if $bvf->class_SO_term(undef, 1) ne "copy_number_variation";
+
+  my $support_vars  = $bvf->structural_variation->get_all_SupportingStructuralVariants;
```
Comment (dglemos) [inline modules/Bio/EnsEMBL/Variation/Utils/VariationEffect.pm:163]: Same here.

Comment (dglemos) [inline modules/Bio/EnsEMBL/Variation/Utils/VariationEffect.pm:163]: Same here.

**Review by dglemos — APPROVED** [review]

**Comment by nuno-agostinho** [comment]
Backported to `release/110` and `postreleasefix/110`.

**Pull request merged**